ServerCore/ServerMessages/PresenceStore.cs

```diff
@@ -0,0 +1,122 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using ServerCore.Pages.Components;
+
+namespace ServerCore.ServerMessages
+{
+    /// <summary>
+    /// Stores player presence on team puzzle pages and notifies clients when it changes
+    /// </summary>
+    public class PresenceStore
+    {
+        ConcurrentDictionary<int, TeamStore> TeamStores { get; } = new ConcurrentDictionary<int, TeamStore>();
+
+        public PresenceStore(ServerMessageListener messageListener)
+        {
+            messageListener.OnPresence += OnPresenceChange;
+        }
+
+        public TeamPuzzleStore GetOrCreateTeamPuzzleStore(int teamId, int puzzleId)
+        {
+            TeamStore teamStore = new TeamStore();
+            if (!TeamStores.TryAdd(teamId, teamStore))
+            {
+                teamStore = TeamStores[teamId];
+            }
+
+            TeamPuzzleStore teamPuzzleStore = new TeamPuzzleStore();
+            if (!teamStore.TeamPuzzleStores.TryAdd(puzzleId, teamPuzzleStore))
+            {
+                teamPuzzleStore = teamStore.TeamPuzzleStores[puzzleId];
+            }
+
+            return teamPuzzleStore;
+        }
+
+        /// <summary>
+        /// Returns a list of all present users on puzzles
+        /// </summary>
+        public List<PresenceMessage> GetAllPresence()
```

Comment:
I'm having trouble walking through all the logic here, but I'm wondering if there's a way for there to be more presence messages acquired between the time that a call to GetAllPresence() starts and when the SignalR hub knows to broadcast new presence messages to the new server.

Reply:
Like, is there any sort of lock required, or is the idea that the listener is acquired before the list is sent, so there may just be some duplicates?

Reply:
The ConcurrentDictionaries are threadsafe, so they won't crash without locks. The question is whether a presence change received during this process and missed by enumeration (or later like while the reply is in transit) would cause a problem. I think the answer is mostly no -- by the time an instance can request presence, it's listening for presence messages, so the new instance doesn't need the existing instance to forward them. The case that I'm not sure how to handle (but hope is minimal) is this can cause reordering.

Imagine pageInstance 123 is present when server 2 spins up. It calls GetAllPresence on server 1, which adds pageInstance 123 to the list. Right after that, pageInstance 123 leaves and sends a message that it's gone. Both servers receive the message. Server 1 removes 123 from its PresenceStore. Server 2's PresenceStore is empty and nothing changes. Then Server 2 receives AllPresence from Server 1, which still has 123 on the list. Server 2 puts 123 in the Store and it never goes away.

The only ideas I have to fully prevent that would be either using a real database or figuring out a cross-machine locking protocol, but I'm not sure that's justified for the case above. What do you think?